File: features/merchant/gold_merchant_common/src/main/java/com/tokopedia/gm/common/data/source/GMCommonDataSource.java
package com.tokopedia.gm.common.data.source;

import com.tokopedia.gm.common.data.source.cloud.GMCommonCloudDataSource;
import com.tokopedia.gm.common.data.source.cloud.model.RequestCashbackModel;
import com.tokopedia.network.mapper.DataResponseMapper;

import javax.inject.Inject;

import rx.Observable;
import rx.functions.Func1;

/**
 * @author sebastianuskh on 3/8/17.
 */
public class GMCommonDataSource {
    private GMCommonCloudDataSource gmCommonCloudDataSource;

    @Inject
    public GMCommonDataSource(GMCommonCloudDataSource gmCommonCloudDataSource) {
        this.gmCommonCloudDataSource = gmCommonCloudDataSource;
    }

    public Observable<Boolean> setCashback(String productId, int cashback) {
        return gmCommonCloudDataSource.setCashback(new RequestCashbackModel(Long.parseLong(productId), cashback))
                .map(new DataResponseMapper<String>())
                .map(new Func1<String, Boolean>() {
                    @Override
                    public Boolean call(String s) {
                        return s != null;
                    }
                });
    }

}

File: features/merchant/gold_merchant_common/src/main/java/com/tokopedia/gm/common/data/source/GMCommonDataSource.kt
package com.tokopedia.gm.common.data.source

import javax.inject.Inject
import com.tokopedia.gm.common.data.source.cloud.GMCommonCloudDataSource
import com.tokopedia.gm.common.data.source.cloud.model.RequestCashbackModel
import com.tokopedia.network.mapper.DataResponseMapper
import rx.Observable

/**
 * @author sebastianuskh on 3/8/17.
 */
class GMCommonDataSource @Inject constructor(private val gmCommonCloudDataSource: GMCommonCloudDataSource) {
    fun setCashback(productId: String, cashback: Int): Observable<Boolean?> {
        return gmCommonCloudDataSource.setCashback(
            RequestCashbackModel(
                productId.toLong(),
                cashback
            )
        )
            .map(DataResponseMapper())
            .map { s -> s != null }
    }
}